server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,915 @@
+const MAX_PROCESSES = 1;
+
+interface DatabaseRequest {
+	statement: string;
+	type: 'all' | 'get' | 'run' | 'transaction';
+	data: AnyObject | any[];
+}
+
+interface DatabaseResult {
+	/** if there's an error, this is it. */
+	error?: string;
+	result?: any;
+}
+
+const STATUS_COLORS: {[k: string]: string} = {
+	idle: '#ff7000',
+	online: '#009900',
+	busy: '#cc3838',
+};
+
+const STATUS_TITLES: {[k: string]: string} = {
+	online: 'Online',
+	idle: 'Idle',
+	busy: 'Busy',
+	offline: 'Offline',
+};
+
+export const Friends = new class {
+	/** `Map<oldID, newID> */
+	renames: Map<string, string>;
+	database?: Database;
+	/** `Map<transferTo, transferFrom>` */
+	transferRequests: Map<string, string>;
+	constructor() {
+		this.setupDatabase();
+		this.renames = this.getRenames();
+		const existingRequests = 'Chat' in global ? Chat.oldPlugins.friends?.transferRequests : undefined;
+		this.transferRequests = existingRequests || new Map();
+		void this.checkExpiringRequests();
+	}
+	setupDatabase() {
+		if (!Config.usesqlite) return;
+		const Sqlite = require('better-sqlite3');
+		try {
+			this.database = new Sqlite(`${__dirname}/../../databases/friends.db`, {fileMustExist: true});
+		} catch (e) {
+			this.database = new Sqlite(`${__dirname}/../../databases/friends.db`);
+			this.database!.exec(FS('databases/schemas/friends.sql').readSync());
+		}
+	}
+	async checkExpiringRequests() {
+		if (!this.database) return;
+		// requests expire after one month. this is checked both on intialization
+		// (hotpatch, to ensure accuracy in case they both don't log in)
+		// and when the user the request is sent to / sent from logs in.
+		const results = this.database.prepare(`SELECT * FROM friend_requests`).all();
+		const removed = [];
+		for (const request of results) {
+			if ((Date.now() - request.sent_at) > REQUEST_EXPIRY_TIME) {
+				// expires after a month
+				await this.removeRequest(request.receiver, request.sender);
+				removed.push(request);
+			}
+		}
+		return removed;
+	}
+	async getFriends(user: User) {
+		const results = await this.query({
+			statement: 'get', type: 'all', data: [user.id, MAX_FRIENDS],
+		});
+		const friends: Set<string> = new Set(results.map((item: AnyObject) => item.friend));
+		return friends;
+	}
+	async getRequests(user: User) {
+		const sent: Set<string> = new Set();
+		const received: Set<string> = new Set();
+		if (user.settings.blockFriendRequests) {
+			// delete any pending requests that may have been sent to them while offline and return
+			await this.query({
+				statement: 'deleteRequest', type: 'run', data: [user.id],
+			});
+			return {sent, received: null};
+		}
+		const sentResults = await this.query({
+			statement: 'getSent', type: 'all', data: [user.id],
+		});
+		for (const request of sentResults) {
+			if ((Date.now() - request.sent_at) > (30 * 24 * 60 * 60 * 1000)) {
+				// expires after a month
+				await this.removeRequest(request.receiver, request.sender);
```

Comment:
`O(n)` queries, correlation, etc.

Honestly, you might just want some sort of cron service in PS to do tasks like this.